### PR TITLE
Finish and test FaradayMiddleware implementation

### PR DESCRIPTION
**Pull request by dazuma**
Fixes https://github.com/census-instrumentation/opencensus-ruby/issues/12

### CODE REVIEW

File: lib/opencensus/trace/integrations/faraday_middleware.rb [hunk 1]
```diff
@@ -17,53 +17,87 @@
...
+          span.put_attribute "/rpc/request/size", body_size if body_size
 
-          # Only sets request size if request is not sent yet.
-          unless env.status
-            request_body = env.body || ""
-            labels["/rpc/request/size"] = request_body.bytesize.to_s
-          end
+          trace_context = span.context.to_trace_context_header
+          headers = env[:request_headers] ||= {}
+          headers["Trace-Context"] = trace_context
```
Comment (chingor13) [inline lib/opencensus/trace/integrations/faraday_middleware.rb]: This header varies depending on the propagation formatter. i.e. `X-Cloud-Trace` for the `CloudTrace` formatter.

We may need to rethink how to make this configurable, but can probably refactor later.

File: lib/opencensus/trace/integrations/faraday_middleware.rb [hunk 2]
```diff
@@ -17,53 +17,87 @@
...
+        def initialize app, span_context: nil, span_name: nil, sampler: nil
+          @app = app
+          @span_context = span_context || OpenCensus::Trace
+          @span_name = span_name || DEFAULT_SPAN_NAME
+          @sampler = sampler
+        end
 
-            response
+        ##
+        # Wraps an HTTP call with a span with the request/response info.
```
Comment (chingor13) [inline lib/opencensus/trace/integrations/faraday_middleware.rb]: This one is missing the YARD param definition or should be marked as internal.

File: lib/opencensus/trace/integrations/faraday_middleware.rb [hunk 3]
```diff
@@ -17,53 +17,87 @@
...
-            response = @app.call env
+        ## The default name for Faraday spans
+        DEFAULT_SPAN_NAME = "faraday_request".freeze
 
-            add_response_labels span, env if span
+        ##
+        # Create a FaradayMiddleware.
+        #
+        # @param [#call] app Next item on the middleware stack.
+        # @param [SpanContext] span_context The span context within which
```
Comment (chingor13) [inline lib/opencensus/trace/integrations/faraday_middleware.rb]: If we're letting a user provide an `SpanContext` instance at initialization, we might want to let them specify on a request time via the request context.

File: lib/opencensus/trace/integrations/faraday_middleware.rb (hunk 2 above)
Comment (dazuma) [reply lib/opencensus/trace/integrations/faraday_middleware.rb]: Fixed

File: lib/opencensus/trace/integrations/faraday_middleware.rb (hunk 3 above)
Comment (dazuma) [reply lib/opencensus/trace/integrations/faraday_middleware.rb]: Agreed in principle, but I'm not sure what that looks like in Faraday. The [request class](https://github.com/lostisland/faraday/blob/master/lib/faraday/request.rb) doesn't seem to be extensible.

File: lib/opencensus/trace/integrations/faraday_middleware.rb (hunk 1 above)
Comment (dazuma) [reply lib/opencensus/trace/integrations/faraday_middleware.rb]: True. But I'll defer this to https://github.com/census-instrumentation/opencensus-ruby/issues/29

File: lib/opencensus/trace/integrations/faraday_middleware.rb (hunk 3 above)
Comment (chingor13) [reply lib/opencensus/trace/integrations/faraday_middleware.rb]: I was looking at the last paragraph of https://github.com/lostisland/faraday#basic-use

Comment (dazuma) [reply lib/opencensus/trace/integrations/faraday_middleware.rb]: Ick. But I guess it works. Added support for it.

**Comment by dazuma** [comment]
I also expanded the class docs.

**Pull request merged**